File: src/geom.py
#!/usr/bin/env python

from numpy import sqrt


def det2(ab, cd):
    (a, b) = ab
    (c, d) = cd
    return (a * d) - (b * c)


def det3(abc, def_, ghi):
    (a, b, c) = abc
    (d, e, f) = def_
    (g, h, i) = ghi
    return (a * det2((e, f), (h, i))) \
        - (b * det2((d, f), (g, i))) \
        + (c * det2((d, e), (g, h)))


def det4(abcd, efgh, ijkl, mnop):
    (a, b, c, d) = abcd
    (e, f, g, h) = efgh
    (i, j, k, l) = ijkl
    (m, n, o, p) = mnop
    return (a * det3((f, g, h), (j, k, l), (n, o, p))) \
        - (b * det3((e, g, h), (i, k, l), (m, o, p))) \
        + (c * det3((e, f, h), (i, j, l), (m, n, p))) \
        - (d * det3((e, f, g), (i, j, k), (m, n, o)))


def ccw(a, b, c):
    (ax, ay) = a
    (bx, by) = b
    (cx, cy) = c
    return ((cy - ay) * (bx - ax)) > ((by - ay) * (cx - ax))


def intersect(ab, cd):
    (a, b) = ab
    (c, d) = cd
    return (ccw(a, c, d) != ccw(b, c, d)) and (ccw(a, b, c) != ccw(a, b, d))


def slope_intercept(a, b):
    (ax, ay) = a
    (bx, by) = b
    if ax == bx:
        return (None, None)
    else:
        m = (by - ay) / (bx - ax)
        b = ay - (m * ax)
        return (m, b)


def point_of_intersection(ab, cd):
    if intersect(ab, cd):
        ((ax, ay), (bx, by)) = ab
        ((cx, cy), (dx, dy)) = cd
        xdelta = (ax - bx, cx - dx)
        ydelta = (ay - by, cy - dy)
        denominator = det2(xdelta, ydelta)
        if denominator != 0:
            d = (det2(*ab), det2(*cd))
            x = det2(d, xdelta) / denominator
            y = det2(d, ydelta) / denominator
            return (x, y)
    return None


def circle_of_points(a, b, c):
    (ax, ay) = a
    (bx, by) = b
    (cx, cy) = c
    if (a == b) or (b == c) or (a == c) or ((ax == bx) and (bx == cx)) \
            or ((ay == by) and (by == cy)):
        return None
    else:
        ax2 = ax * ax
        ay2 = ay * ay
        bx2 = bx * bx
        by2 = by * by
        cx2 = cx * cx
        cy2 = cy * cy
        axy2 = ax2 + ay2
        bxy2 = bx2 + by2
        cxy2 = cx2 + cy2
        A2 = 2 * det3((ax, ay, 1), (bx, by, 1), (cx, cy, 1))
        B = det3((axy2, ay, 1), (bxy2, by, 1), (cxy2, cy, 1))
        C = det3((axy2, ax, 1), (bxy2, bx, 1), (cxy2, cx, 1))
        x = (B / A2)
        y = -(C / A2)
        a = (x - ax)
        b = (y - ay)
        r = sqrt((a * a) + (b * b))
        return ((x, y), r)


def point_in_circle(a, b, c, d):
    (ax, ay) = a
    (bx, by) = b
    (cx, cy) = c
    (dx, dy) = d
    ax2 = ax * ax
    ay2 = ay * ay
    bx2 = bx * bx
    by2 = by * by
    cx2 = cx * cx
    cy2 = cy * cy
    dx2 = dx * dx
    dy2 = dy * dy
    D = det4(
        (ax, ay, (ax2 + ay2), 1),
        (bx, by, (bx2 + by2), 1),
        (cx, cy, (cx2 + cy2), 1),
        (dx, dy, (dx2 + dy2), 1),
    )
    # D <  0  ->  D is within circle(A, B, C)
    # D == 0  ->  D is co-circular with A, B, C
    # D  > 0  ->  D is outside circle(A, B, C)
    return D

File: go/src/geom/geom.go
package geom

import (
    "fmt"
    "math"
)

type Pair struct {
    X, Y float64
}

type Segment struct {
    A, B Pair
}

type Triple struct {
    X, Y, Z float64
}

type Quad struct {
    X, Y, Z, W float64
}

type Circle struct {
    Center Pair
    Radius float64
}

func det2(a Pair, b Pair) float64 {
    return (a.X * b.Y) - (a.Y * b.X)
}

func det3(a, b, c Triple) float64 {
    return (a.X * det2(Pair{b.Y, b.Z}, Pair{c.Y, c.Z})) -
        (a.Y * det2(Pair{b.X, b.Z}, Pair{c.X, c.Z})) +
        (a.Z * det2(Pair{b.X, b.Y}, Pair{c.X, c.Y}))
}

func det4(a, b, c, d Quad) float64 {
    A := a.X * det3(
        Triple{b.Y, b.Z, b.W},
        Triple{c.Y, c.Z, c.W},
        Triple{d.Y, d.Z, d.W},
    )
    B := a.Y * det3(
        Triple{b.X, b.Z, b.W},
        Triple{c.X, c.Z, c.W},
        Triple{d.X, d.Z, d.W},
    )
    C := a.Z * det3(
        Triple{b.X, b.Y, b.W},
        Triple{c.X, c.Y, c.W},
        Triple{d.X, d.Y, d.W},
    )
    D := a.W * det3(
        Triple{b.X, b.Y, b.Z},
        Triple{c.X, c.Y, c.Z},
        Triple{d.X, d.Y, d.Z},
    )
    return A - B + C - D
}

func Ccw(a, b, c Pair) bool {
    return ((c.Y - a.Y) * (b.X - a.X)) > ((b.Y - a.Y) * (c.X - a.X))
}

/* https://en.wikipedia.org/wiki/Line%E2%80%93line_intersection */
func PointOfIntersection(a, b Segment) (Pair, error) {
    x1 := a.A.X
    x2 := a.B.X
    x3 := b.A.X
    x4 := b.B.X
    y1 := a.A.Y
    y2 := a.B.Y
    y3 := b.A.Y
    y4 := b.B.Y
    denominator := ((x1 - x2) * (y3 - y4)) - ((y1 - y2) * (x3 - x4))
    if denominator != 0.0 {
        t := (((x1 - x3) * (y3 - y4)) - ((y1 - y3) * (x3 - x4))) / denominator
        u := -(((x1 - x2) * (y1 - y3)) - ((y1 - y2) * (x1 - x3))) / denominator
        if 0.0 <= t && t <= 1.0 && 0.0 <= u && u <= 1.0 {
            x := x1 + (t * (x2 - x1))
            y := y1 + (t * (y2 - y1))
            return Pair{x, y}, nil
        }
    }
    return Pair{}, fmt.Errorf("PointOfIntersection(%v, %v)", a, b)
}

func SlopeIntercept(a Segment) (float64, float64, error) {
    if a.A.X == a.B.X {
        return 0.0, 0.0, fmt.Errorf("SlopeIntercept(%v)", a)
    }
    m := (a.B.Y - a.A.Y) / (a.B.X - a.A.X)
    b := a.A.Y - (m * a.A.X)
    return m, b, nil
}

func CircleOfPoints(a, b, c Pair) (Circle, error) {
    if (a == b) || (b == c) || (a == c) ||
        ((a.X == b.X) && (b.X == c.X)) ||
        ((a.Y == b.Y) && (b.Y == c.Y)) {
        return Circle{}, fmt.Errorf("CircleOfPoints(%v, %v, %v)", a, b, c)
    }
    ax2 := a.X * a.X
    ay2 := a.Y * a.Y
    bx2 := b.X * b.X
    by2 := b.Y * b.Y
    cx2 := c.X * c.X
    cy2 := c.Y * c.Y
    axy2 := ax2 + ay2
    bxy2 := bx2 + by2
    cxy2 := cx2 + cy2
    A2 := 2.0 * det3(
        Triple{a.X, a.Y, 1.0},
        Triple{b.X, b.Y, 1.0},
        Triple{c.X, c.Y, 1.0},
    )
    B := det3(
        Triple{axy2, a.Y, 1.0},
        Triple{bxy2, b.Y, 1.0},
        Triple{cxy2, c.Y, 1.0},
    )
    C := det3(
        Triple{axy2, a.X, 1.0},
        Triple{bxy2, b.X, 1.0},
        Triple{cxy2, c.X, 1.0},
    )
    x := B / A2
    y := -(C / A2)
    r := math.Sqrt(((x - a.X) * (x - a.X)) + ((y - a.Y) * (y - a.Y)))
    return Circle{Pair{x, y}, r}, nil
}

func PointInCircle(a, b, c, d Pair) (float64, error) {
    if (a == b) || (b == c) || (a == c) ||
        ((a.X == b.X) && (b.X == c.X)) ||
        ((a.Y == b.Y) && (b.Y == c.Y)) {
        return 0.0, fmt.Errorf("PointInCircle(%v, %v, %v, %v)", a, b, c, d)
    }
    // det4(...) <  0   ->  d is within circle(a, b, c)
    // det4(...) == 0   ->  d is co-circular with a, b, c
    // det4(...)  > 0   ->  d is outside circle(a, b, c)
    return det4(
        Quad{a.X, a.Y, (a.X * a.X) + (a.Y * a.Y), 1.0},
        Quad{b.X, b.Y, (b.X * b.X) + (b.Y * b.Y), 1.0},
        Quad{c.X, c.Y, (c.X * c.X) + (c.Y * c.Y), 1.0},
        Quad{d.X, d.Y, (d.X * d.X) + (d.Y * d.Y), 1.0},
    ), nil
}
